clients/web/src/collections/Collection.js

```diff
@@ -87,46 +129,138 @@ var Collection = Backbone.Collection.extend({
+                this.pageOffsetStack = [];
+            }
             this.offset = 0;
         } else {
             this.params = params || {};
         }
 
+        if (this.filterFunc && !this.append) {
+            this.pageOffsetStack.push(this.offset);
+        }
+
         var limit = this.pageLimit > 0 ? this.pageLimit + 1 : 0;
 
-        var xhr = restRequest({
-            path: this.altUrl || this.resourceName,
-            data: _.extend({
-                limit: limit,
-                offset: this.offset,
-                sort: this.sortField,
-                sortdir: this.sortDir
-            }, this.params)
-        }).done(_.bind(function (list) {
-            if (this.pageLimit > 0 && list.length > this.pageLimit) {
-                // This means we have more pages to display still. Pop off
-                // the extra that we fetched.
-                list.pop();
-                this._hasMorePages = true;
-            } else {
-                this._hasMorePages = false;
-            }
+        var finalList = []; /* will be built up in pieces */
 
-            this.offset += list.length;
+        function fetchListFragment() {
+            var xhr = restRequest({
+                path: this.altUrl || this.resourceName,
+                data: _.extend({
+                    limit: limit,
+                    offset: this.offset,
+                    sort: this.sortField,
+                    sortdir: this.sortDir
+                }, this.params)
+            });
 
-            if (list.length > 0 || reset) {
-                if (this.append && !reset) {
-                    this.add(list);
+            var result = $.when(xhr).then(_.bind(function (list) {
+                if (!_.isArray(list)) {
+                    list = [list];
+                }
+
+                if (this.pageLimit > 0 && list.length > this.pageLimit) {
+                    // This means we have more pages to display still. Pop off
+                    // the extra that we fetched.
+                    list.pop();
+                    this._hasMorePages = true;
                 } else {
-                    this.reset(list);
+                    this._hasMorePages = false;
                 }
-            }
 
-            this.trigger('g:changed');
-        }, this));
-        xhr.girder = {fetch: true};
-        return xhr;
+                var offsetDelta = list.length;
+
+                /*
+                 * If filtering, decorate the list with their pre-filtered
+                 * indexes.  The index will be needed when adjusting the offset.
+                 */
+                if (this.filterFunc) {
+                    var filter = this.filterFunc;
+                    list = (
+                        list
+                            .map(function (x, index) { return [index, x]; })
+                            .filter(function (tuple) {
+                                return filter(tuple[1]);
+                            })
```

Comment:
Alright, I'll add another variable.  If you absolutely cannot *stand* to have an overloaded variable name. :grin:

I don't know what you mean by "standard JS api" in this context; AFAIK, my usage is standard while the line of code in your comment is not?

edit: Oh, I think I see what you were getting at (maybe ;P).  No, the filter function provided does not conform to the standard Array.filter callback signature and that is intentional.  The filter function has online semantics, which is the major characteristic that sets it apart from a standard Array.filter callback.  The filter is applied to data as it is fetched, so it doesn't even make sense to consider the collection nor any internal index on the temporarily fetched data as part of the filtering criteria (the former because the collection may be incomplete, and the latter because the index has no meaning outside of the fetch and is subject to change from one fetch to the next).